### PR TITLE
UKSW-10 Unit tests for visibility receive (continued)

### PR DESCRIPTION
**Pull request by njthykkathu**
<!—UKSW-10 Unit tests for visibility receive (continued) -->

## Description:
A test developed to confirm the behaviour of handling SPEAD heaps by generating in memory SPEAD heaps and calling the function that decodes the header + payload from the memory buffer. 

Note. Changes have been made to stream.c file for visibility baseline count (line 149). When testing it was found that num_baselines is getting converted using be32toh. We don’t see a reason for this conversion as the items are already getting converted using be64toh in line 96. It is the same for visibility time fraction and visibility timestamp count but has not those are not updated.

## Testing instructions:
Build Instructions
To build the code on a local machine, ensure make and CMake are both installed and give the following commands from the current directory:
```
mkdir build
cd build
cmake ..
make
```
To run the unit tests from the build directory, either run
`ctest`
or run the unit test binary directly using:
`./test/recv_test`

## Types of changes

- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] The code follows the code style of this project.
- [ ] The changes require a documentation update.
- [ ] Documentation has been updated accordingly.
- [X] Tests cover all changes in this PR.
- [X] All new and existing tests passed.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -80,4 +80,4 @@ jobs:
     install:
     - pip install -r testing_requirements.txt
     script: ./tools/run_tests.sh sip/platform/logging
-    after_success: codecov
```
Comment (fdulwich) [inline .travis.yml]: (Not sure why this change was committed?)

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp [hunk 2]
```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
...
+
+    // get start of heap
+    heap_start = buf + 8 + num_items * sizeof(uint64_t);
+
+    // insert scan_id at heap_start, 8 bytes
+    scan_ptr = (uint64_t *) heap_start;
+    *scan_ptr = scan_id;
+    *scan_ptr = htobe64(scan_id);
+    ++scan_ptr;
+    vis_ptr = (struct DataType *) scan_ptr;
```
Comment (fdulwich) [inline sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: I think it would be a bit clearer to write this instead, and remove the line above it:
```suggestion
    struct DataType* vis_ptr = (struct DataType*) (scan_ptr + 1);
```
The separate declaration of `vis_ptr` on line 140 is then also not required.

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp [hunk 3]
```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
...
+        const uint64_t item_id = (uint64_t) p_items[i].id;
+        const uint64_t item_addr_or_val = (uint64_t) p_items[i].val;
+	    const uint64_t imm_or_abs = p_items[i].imm << 63;
+        //printf("Item Value: %i\n", item_val[i]);
+        const uint64_t item =
+                imm_or_abs |
+                (item_id << heap_address_bits) |
+                item_addr_or_val;
+        items[i] = htobe64(item);
+	switch (p_items[i].id)
```
[the remark this answers is not in the document]
Comment (apmcd42) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: The whole point of the switch there was to pull out values to be used in the tests further on. Should we be testing the heap counter? Are there other values we should be pulling out for testing? If not, then this switch statement can be removed.

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp [hunk 4]
```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
...
+    ASSERT_TRUE(receiver != NULL);
+
+    // Passing the buffer and stream to stream decode
+    stream_decode(receiver->streams[0], buf, 0);
+
+    // Testing
+    ASSERT_TRUE(receiver->streams[0] != NULL);
+    ASSERT_TRUE(receiver->buffers[0] != NULL);
+    ASSERT_TRUE(receiver->num_buffers != 0);
+    ASSERT_TRUE(num_baselines == receiver->buffers[0]->num_baselines);
```
[the remark this answers is not in the document]
Comment (apmcd42) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: Okay. Is there any documentation on the google tests that I can read up on?

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp [hunk 5]
```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
...
+    ASSERT_TRUE(vis_ptr->fd == receiver->buffers[0]->vis_data->fd);
+    ASSERT_TRUE(vis_ptr->tci == receiver->buffers[0]->vis_data->tci);
+
```
[the remark this answers is not in the document]
Comment (apmcd42) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: I'll look into that.

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (fdulwich) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: There's a good introduction here, and the linked section explains the difference between ASSERT* and EXPECT*: https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp [hunk 6]
```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
...
+	}
+    }
+
+    // get start of heap
+    heap_start = buf + 8 + num_items * sizeof(uint64_t);
+
+    // insert scan_id at heap_start, 8 bytes
+    scan_ptr = (uint64_t *) heap_start;
+    *scan_ptr = scan_id;
+    *scan_ptr = htobe64(scan_id);
```
[the remark this answers is not in the document]
Comment (apmcd42) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: That first line should be deleted anyway

[the remark this answers is not in the document]
Comment (fdulwich) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: I thought `scan_ptr` didn't need its endianness swapped, as it's not used in immediate addressing mode, so the first line would be correct in that case?

File: sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp (hunk 2 above)
Comment (apmcd42) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: I have to admit a preference to putting all variable declarations at the start of a code block. We may be C programmers, but that doesn't make us anarchists :-)

Comment (fdulwich) [reply sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp]: My motivation for flagging this was because it's logically wrong to increment a variable called `scan_ptr`, which then makes it no longer point to the scan ID. Code like this is unsafe, as any future use of `scan_ptr` would then be an error.
